File: algorithms/easy/count_vowel_substrings_of_a_string.ex
defmodule Solution do
  @spec count_vowel_substrings(word :: String.t()) :: integer
  def count_vowel_substrings(word) do
    vowel_index = %{"a" => -1, "e" => -1, "i" => -1, "o" => -1, "u" => -1}
    word |> String.graphemes() |> aux(0, -1, vowel_index, 0)
  end

  def aux([], _idx, _last_consonant, _vowel_index, ans), do: ans

  def aux([c | rest], idx, last_consonant, vowel_index, ans) do
    if Map.has_key?(vowel_index, c) do
      vowel_index = Map.put(vowel_index, c, idx)
      the_most_left_vowel = vowel_index |> Map.values() |> Enum.min()
      ans = ans + max(0, the_most_left_vowel - last_consonant)
      aux(rest, idx + 1, last_consonant, vowel_index, ans)
    else
      aux(rest, idx + 1, idx, vowel_index, ans)
    end
  end
end

File: algorithms/easy/count_vowel_substrings_of_a_string.rs
use std::collections::HashMap;

impl Solution {
    pub fn count_vowel_substrings(word: String) -> i32 {
        let mut ans = 0;

        let word: Vec<_> = word.chars().collect();

        let mut m = HashMap::new();
        m.insert('a', -1);
        m.insert('e', -1);
        m.insert('i', -1);
        m.insert('o', -1);
        m.insert('u', -1);

        let mut last_consonant = -1;

        for i in 0..word.len() {
            if let Some(n) = m.get_mut(&word[i]) {
                *n = i as i32;
                let the_most_left_vowel: i32 = *m.values().min().unwrap();
                ans += (the_most_left_vowel - last_consonant).max(0);
            } else {
                last_consonant = i as i32;
            }
        }
        ans
    }
}
